sonar-batch/src/main/java/org/sonar/batch/bootstrap/BatchWsClient.java

```diff
@@ -0,0 +1,130 @@
+ * Copyright (C) 2008-2014 SonarSource
+ * mailto:contact AT sonarsource DOT com
+ *
+ * SonarQube is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * SonarQube is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.batch.bootstrap;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Joiner;
+import com.google.gson.JsonArray;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+import java.net.HttpURLConnection;
+import java.util.ArrayList;
+import java.util.List;
+import javax.annotation.Nullable;
+import org.apache.commons.lang.StringUtils;
+import org.sonar.api.CoreProperties;
+import org.sonar.api.utils.MessageException;
+import org.sonar.api.utils.log.Logger;
+import org.sonar.api.utils.log.Loggers;
+import org.sonar.api.utils.log.Profiler;
+import org.sonarqube.ws.client.HttpException;
+import org.sonarqube.ws.client.WsClient;
+import org.sonarqube.ws.client.WsConnector;
+import org.sonarqube.ws.client.WsRequest;
+import org.sonarqube.ws.client.WsResponse;
+
+import static java.lang.String.format;
+
+public class BatchWsClient {
+
+  private static final Logger LOG = Loggers.get(BatchWsClient.class);
+
+  private final WsClient target;
+  private final boolean hasCredentials;
+  private final String publicBaseUrl;
+
+  public BatchWsClient(WsClient target, boolean hasCredentials, @Nullable String publicBaseUrl) {
+    this.target = target;
+    this.hasCredentials = hasCredentials;
+    if (StringUtils.isBlank(publicBaseUrl)) {
+      this.publicBaseUrl = target.wsConnector().baseUrl();
+    } else {
+      this.publicBaseUrl = publicBaseUrl.replaceAll("(/)+$", "") + "/";
+    }
+  }
+
+  /**
+   * @throws IllegalStateException if the request could not be executed due to
+   *     a connectivity problem or timeout. Because networks can
+   *     fail during an exchange, it is possible that the remote server
+   *     accepted the request before the failure
+   * @throws HttpException if the response code is not in range [200..300)
+   */
+  public WsResponse call(WsRequest request) {
+    Profiler profiler = Profiler.createIfDebug(LOG).start();
+    WsResponse response = target.wsConnector().call(request);
+    profiler.stopDebug(format("%s %d %s", request.getMethod(), response.code(), response.requestUrl()));
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) An 'int' is expected rather than a !unknown!. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2275)